Markdown Gutenberg block for Jetpack

Ferdev opened the pull request:
### Changes proposed in this Pull Request:

#### Description

Creation of a Markdown Gutenberg block. The user will be able to enter Markdown code in Gutenberg, and it will be rendered into HTML in the client. The block will allow creating new content using Markdown, as well as editing existing Markdown blocks.

This PR comes from the previous work done by @nuzzio in the PRs #9357 and https://github.com/WordPress/gutenberg/pull/6491 (thanks!) and addresses #9201. As a summary, a decision was made to perform the Markdown to HTML conversion in the client, which is the purpose of this PR.

The block will have the following features:

1. [x] Adds a content-editable element to input Markdown text.
2. [x] Adds a preview zone to display the HTML rendered from the Markdown code.
4. [x] Live preview: entered Markdown code will be rendered as HTML as the user types it.
3. [x] Doesn't load the old classic markdown support code if the Gutenberg plugin is installed and active.
6. [ ] ~~Basic text editing buttons to make text bold and italic, create links and add images. They should mimic the UX in the RichText Gutenberg editor component.~~
7. [ ] ~~Add a checkbox to allow the user disable the live-preview (for very big markdown sources).~~
5. [ ] ~~A Markdown syntax legend in the block description (for Markdown newbies).~~

The last three features have been moved to PR #9825

The component will be developed following the order specified in the list above.

This PR notes will be updated as it evolves.

### Testing instructions:

Steps to test the Markdown block component:

1. Install the Gutenberg plugin in your WordPress environment.
2. Enable "Write posts or pages in plain-text Markdown syntax" in your Jetpack config page.
3. Create a new post, add a Markdown block and write some markdown code.
4. Save a draft of the post, or publish it.
5. The markdown code should be rendered as HTML in the published post/draft.

Posts can be edited as well.

~~The block Javascript code will be unit tested using the updated test setup in the PR #9687 that allows testing of Javascript code within modules.~~

~~The block development will be test driven so the test suite will grow as the development progresses.~~

Unfortunately, until issue https://github.com/WordPress/gutenberg/issues/3955 is addressed, custom blocks developed outside Gutenberg cannot be unit tested due to the dependencies they have with Gutenberg core blocks and components.

### Proposed changelog entry for your changes:

Added a markdown block for the Gutenberg plugin.

### Checklist:

- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards
- [x] My code has proper inline documentation.

ehg commented on modules/markdown/assets/js/jetpack-markdown-block.jsx:
```diff
@@ -0,0 +1,70 @@
+	category: 'formatting',
+
+	attributes: {
+		//The Markdown source is saved in the block content comments delimiter
```
What was your reasoning around saving the source in the attributes? Why not in the block's body? Why not in block meta?

Ferdev replied:
I thought it was the cleanest solution. It makes super simple to link the markdown source with the block's content (in case you add more than one markdown blocks).
When making blocks I guess we should strive to save the attributes in the block's content, that way we avoid data duplication, but I couldn't find a way to store the markdown source in the block's content and then hide it in the published post (but I'm a total WordPress newbie, maybe there's a way)...
> Why not in block meta?
I guess you meant in the post's meta, right? I guess one of the uses of the post's meta is to share post's info in another parts of WordPress: other plugins, themes, etc. I didn't think the markdown source would be needed in any other place. It would make it harder to manage each markdown block source: if we store the attribute as it is, it'll return a strings array. If we choose to store as json, it'd have to serialize/parse it every time we needed. It seemed way more complex to do it this way.

mcsf replied:
At a glance, I agree with keeping the Markdown source as a regular (HTML-comment-borne) attribute. Aside from ease on the development side, it also gets us data deduplication for free (that is, when rendering the front-end, assuming Gutenberg is stripping HTML comments, the visitor only gets the generated HTML source).

For completion:

- Keeping the Markdown source in the block's body could be achieved using old techniques such as using custom script tags like `<script type="text/markdown">` that the browser can ignore. Gutenberg can be told how to source attributes from these. The downside is that this information is wastefully sent to the visitor.

- It's possible to intercept the block's body to decide what gets sent with the rendered page, but that requires making the block dynamic (and having a PHP render callback to render the block on every page load). I don't see how this option could be better than defaulting to HTML comments for the Markdown source, though.

Ferdev replied:
Thanks for the clarification @mcsf ! Especially the last point :+1:

ehg commented on package.json:
```diff
@@ -63,6 +63,7 @@
     "classnames": "2.2.1",
     "click-outside": "^2.0.2",
     "clipboard": "^1.7.1",
+    "commonmark": "^0.28.1",
```
Why `commonmark`?

ehg replied:
Also, do you think we should pin the version here?

Ferdev replied:
> Why `commonmark`?

I thought using a Markdown parser library made more sense than doing it myself: that could be easily a different project since it isn't something trivial to develop.
I used a commonmark implementation for Elixir in the past, I knew it was fast and reliable. Also, I asked in Slack's #javascript channel and was one of the recommendations. This javascript implementation project is active and well documented, so it was my first choice. Now I've found it lacks one feature and I've switched to markdown-it (which is another implementation of the commonmark specification) in my latest work: the parser does not store the delimiters used in the source, which I use for the live preview part of the project.

> Also, do you think we should pin the version here?

I totally missed this one. I usually use npm to manage my node dependencies, and I had the `npm config set save-prefix='~'` enabled in my local npm config since a long time ago. Yarn doesn't seem to use this setting, and it used the `^` to save the package. I like to use `~` because that way only the patch part is updated, which means any bugs would be fixed in further updates. But maybe you have another policy in Automattic and you're more conservative in this aspect (which I totally understand) and you always use fixed versions of packages. Anyway, thanks for pointing this out!

ehg commented on modules/markdown/assets/js/components/markdown-preview.jsx:
```diff
@@ -0,0 +1,32 @@
+} = window.wp.element;
+
+const markdownReader = new commonmark.Parser();
+const markdownWritter = new commonmark.HtmlRenderer();
```
s/Writter/Writer

mcsf commented on modules/markdown/assets/js/jetpack-markdown-block-editor.jsx:
```diff
@@ -0,0 +1,76 @@
+					{ placeholderSource }
+				</p>;
+			}
+			return <MarkdownLivePreview
```
I suggest tweaking the terminology somewhere. The distinction between "Preview" and "Live Preview" can be confusing.

Ferdev replied:
Totally agree. Actually, this was something I was planning to do because indeed the names are very similar and can be confusing.

mcsf commented on modules/markdown/assets/js/jetpack-markdown-block-editor.jsx:
```diff
@@ -0,0 +1,76 @@
+			const placeholderSource = __( 'Write your _Markdown_ **here**...' );
+
+			if ( ! isSelected && this.isEmpty() ) {
+				return <p className={ `${ className }-placeholder` }>
```
```js
return (
    <p>
```

mcsf commented on modules/markdown/assets/js/jetpack-markdown-block-editor.jsx:
```diff
@@ -0,0 +1,76 @@
+
+		return [
+			editorOrPreviewPanel.call( this )
+		];
```
I don't see the need for `editorOrPreviewPanel` :). Is it an artefact of something? Furthermore, the array shouldn't be needed.

Ferdev replied:
Yeah, there's no need for this function anymore. The code inside the `render` function was more complex when the block had the "Editor" and "Preview" panels. But when I removed that feature I forgot to simplify this code and remove `editorOrPreviewPanel` 
I'll fix that :+1:

mcsf commented on modules/markdown/assets/js/components/markdown-live-preview.jsx:
```diff
@@ -0,0 +1,193 @@
+		};
+	}
+
+	render() {
```
Convention has it that `render` should be the last method in the class body. Also, lifecycle methods should precede any custom methods:
1. constructor
2. componentDid… / shouldCompo…
3. foo / bar
4. render

Ferdev replied:
Thanks! :bowing_man: I didn't know what the convention was.

mcsf commented on modules/markdown/assets/js/components/markdown-live-preview.jsx:
```diff
@@ -0,0 +1,193 @@
+		);
+	}
+
+	shouldComponentUpdate( nextProps, nextState ) {
```
Thinking of the future, `getSnapshotBeforeUpdate` will probably be more apt for the caret position saving.

mcsf commented on modules/markdown/assets/js/components/markdown-live-preview.jsx:
```diff
@@ -0,0 +1,193 @@
+
+		const { source } = props;
+
+		setupMarkdownParser();
```
This will be called on each instantiation of a preview — expectably, multiple times over the lifespan of the editor — thus setting up the same global every time. Seems we should avoid that.

Ferdev replied:
Yep, good catch, thanks! I'll fix that.

mcsf commented on modules/markdown/assets/js/components/markdown-live-preview.jsx:
```diff
@@ -0,0 +1,193 @@
+
+const emptyState = '<p></p>';
+
+const emitChange = function( evt ) {
```
What was the rationale for keeping this function outside the component class?

mcsf replied:
Independently of that, why isn't `emitChange` bound in the class constructor?

mcsf replied:
Yet another question :) : why the name `emitChange`? To me it reads as if the function is doing more than it is, especially compared with some similar handlers found in core Gutenberg.

Ferdev replied:
Answering your first two questions: I guess this was just personal preference. I tend to not expose those functions that aren't supposed to be consumed out of the class. I don't know what is the convention you'd use here, but this was my reasoning.
And regarding the name... Yeah, probably that wasn't my best choice :sweat_smile: I'll give it another shot.

mcsf commented on modules/markdown/assets/js/components/markdown-live-preview.jsx:
```diff
@@ -0,0 +1,193 @@
+
+const ignoreLastInput = function( source ) {
+	return endsWith( source, ' ' ) || endsWith( source, ' ' );
+};
```
Can we represent the space characters differently to make it obvious what they are?

Ferdev replied:
Yeah, totally, this is something I had in my list of improvements. I was planning to create some constants using `String.fromCharCode(0x00A0)` to avoid mistakes between the two different space chars.

Ferdev commented on modules/markdown/assets/js/jetpack-markdown-block.jsx:
```diff
@@ -0,0 +1,70 @@
+	description: [
+		__( 'Write your content in plain-text Markdown syntax.' ),
+		(
+		<p>
```
Ops, my fault, first time using JSX. I'll fix that

Ferdev commented:
Thanks for your time in the review, @mcsf ! Just answered your questions, I'll push some commits tomorrow to tackle your suggestions & improvements.

ehg commented:
How is the live preview functionality meant to behave when I copy & paste MD into the block? I tried copying https://raw.githubusercontent.com/mxstbr/markdown-test-file/master/TEST.md into the block, and it didn't update until I pressed a key.

Also, what subset of MD are we supporting? I can see some oddities when pasting `TEST.md` in :)

Finally, what is the performance impact of live preview? What happens if we have lots of text?

Ferdev commented:
The problem you had pasting markdown code is a bug and it's weird because this was working... I'll perform a git bisect to check where the issue was introduced and, of course, fix it. It's supposed to live preview as you paste any markdown code.

Yeah, I see those oddities too! The parser supports the [Commonmark specification](https://spec.commonmark.org/current/) and the problem has something to do with how the text is pasted in the component and maybe some white-space issues too... I'll check that and will come back with an update.

From the tests I've done, the performance is pretty good even for large texts in Chrome and Firefox, but in the list of pending things I have to perform proper performance tests. I was planning to do that when all features and bugs had been addressed.

Ferdev commented:
I've fixed the issues you had pasting Markdown code, @ehg. This has been addressed in ec681b367c114691617d65bd86a05fb03e69ce1e, 1b4022c6de5e1eb7e60f610d7a1a388e1a4076e7 and 30c51379a59b6ccff6d49a2b5e71e267f529f704. In the description of those commits you'll discover the reasons causing that problem.
I'd like your opinion on the first two commits. I've refactored the code and extracted all the Markdown conversion functionality to an independent library, but I don't like using source code from the original library in our implementation. There's no other way to do it, and actually [this is the recommended way](https://github.com/markdown-it/markdown-it/issues/46) of overriding the library, but still... I'd reimplement the live-preview parser with a custom one

ehg commented:
> I'd like your opinion on the first two commits. I've refactored the code and extracted all the Markdown conversion functionality to an independent library, but I don't like using source code from the original library in our implementation. 

Yeah, this isn't ideal. Do you know why `markdown-it` is trimming spaces? Should this issue be raised upstream?

ehg commented on modules/markdown/assets/js/components/markdown-live-preview.jsx:
```diff
@@ -0,0 +1,164 @@
+	render() {
+		const { ...props } = this.props;
+
+		return createElement(
```
Why are we using `createElement` here instead of JSX?

Ferdev replied:
This was just me using React and JSX for the first time. I'll change it to JSX notation 👍

ehg commented on modules/markdown/assets/js/components/markdown-live-preview.jsx:
```diff
@@ -0,0 +1,164 @@
+			{
+				...props,
+				ref: ( e ) => this.htmlEl = e,
+				onBlur: renderMarkdownPreview.bind( this ),
```
`bind` creates a new function on every render. The `shouldComponentUpdate` looks like it will make this less of a problem, but we should avoid doing this regardless as it can lead to unexpected behaviour.

Ferdev replied:
Yeah, you're totally right. Actually, I found another place where we're creating a new function quite often: [here](https://github.com/Automattic/jetpack/pull/9705/files/30c51379a59b6ccff6d49a2b5e71e267f529f704#diff-26458d622e729f313ea1b570125ab5adR36)
I'll fix both issues 👍

ehg commented on package.json:
```diff
@@ -116,6 +116,7 @@
     "json-loader": "^0.5.7",
     "lodash": "^4.9.0",
     "node-sass": "4.9.0",
+    "markdown-it": "~8.4.1",
```
Let's pin this to the exact version number, in case there's semver principles are skipped upstream :)

ehg commented on webpack.config.js:
```diff
@@ -10,7 +10,8 @@ const webpackConfig = {
 		admin: './_inc/client/admin.js',
-		'static': './_inc/client/static.jsx'
+		'static': './_inc/client/static.jsx',
+		'modules-markdown-block': './modules/markdown/assets/js/jetpack-markdown-block.jsx',
```
Why do we need to add an entry point here? Can we get away without one?

Ferdev replied:
Allow me to investigate this further, please. I saw this was the pattern used in other PRs for Gutenberg blocks in Jetpack and I just thought it was a valid solution.

Ferdev commented:
>> I'd like your opinion on the first two commits. I've refactored the code and extracted all the Markdown conversion functionality to an independent library, but I don't like using source code from the original library in our implementation.

> Yeah, this isn't ideal. Do you know why markdown-it is trimming spaces? Should this issue be raised upstream?

Yeah, it's by design. The CommonMark Spec does not allow heading or trailing whitespace in paragraphs (see point 4.8 paragraphs [here](https://spec.commonmark.org/0.28/)):

> A sequence of non-blank lines that cannot be interpreted as other kinds of blocks forms a paragraph. The contents of the paragraph are the result of parsing the paragraph’s raw content as inlines. The paragraph’s raw content is formed by concatenating the lines and removing initial and final whitespace.

The problem is Markdown-It doesn't offer a setting to customize this, which I think is reasonable since the specification doesn't allow it.

I think our options here are:
1. Create a custom parser for the live preview. Maybe over-kill.
2. Create Markdown-It plugin to disable auto trimming in paragraphs and new lines. More or less what we have now but in a nicer wrapping.
3. Evaluate another Markdown parsers that don't follow the CommonMark spec.

Maybe it's worth to invest more time in a deeper analysis of other options, and then make a final decision. What do you think?

Ferdev commented:
@brentlogan sure! One of the reasons for choosing Markdown-It is it makes easy to extend the specification with custom rules. Actually, the fact that it supports footnotes was a ➕ when I saw the classic editor supported them. Do you think we should implement any other rules not contemplated in the CommonMark spec?

Thanks for your feedback! 🙇

brentlogan commented:
@Ferdev I looked at Markdown Extra to look at its features that I regularly use:

* Footnotes
* Simple table definition
* Using markdown within HTML blocks
* Fenced code blocks

I guess I'm a big fan of Markdown Extra... ;-)

As best I can tell, the latter two are included in Commonmark. Thanks for listening!

Ferdev commented:
yeah @brentlogan , I'd say all your use cases are covered by Markdown-It. You can check its features in their live editor [here](https://markdown-it.github.io/) 🎉

Ferdev commented:
I've performed a more in-depth analysis of existing Javascript Markdown parsers to help us choose the most suited for our needs.

The Markdown block uses the parser in two points: the Save component of the block, and the Live Preview component:

- The Save block receives Markdown source and generates the resulting HTML that will be saved in the post's content.
- The Live Preview component, renders a subset of the Markdown specification (only emphasis, strong, inline code, and headings) as the user types the Markdown source. We need a parser that allows limiting the Markdown features, and also allows tweaking the result to add back the tokens delimiters used by the user.

These are the most used Markdown parsers right now:

### [showdown](https://github.com/showdownjs/showdown)
This is probably the most adopted one. GoogleCloudPlatform, Ghost, Meteor... are some companies using it. Its engine is based in regular expressions and sequential parsing of the Markdown source. This makes difficult to disable parts of the specs, which we need to do for the Live Preview component.
It allows to [create extensions](https://github.com/showdownjs/showdown/wiki/extensions) to extend the functionality of the library.
It has the worst performance out of the 5 in this analysis.

### [commonmark.js](https://github.com/commonmark/commonmark.js)
commonmark.js is a lightweight Markdown parser based in the CommonMark specification. It generates an abstract syntax tree that can be traversed to perform some modification/customization to the resulting HTML. Unfortunately, it lacks some features like storing the token delimiters used by the author of the Markdown source to add them back in the live preview component. It's the fastest of all of them.

### [Marked](https://github.com/markedjs/marked)
Marked is in the analysis just because is one of the most used ones, but it turned out to be not very performant and also it lacks any form of extensibility.

### [markdown-it](https://github.com/markdown-it/markdown-it)
markdown-it is the library being used in the Markdown block currently. The performance is excellent, and it allows easy extensibility, either using the existing [plugins](https://www.npmjs.com/search?q=keywords:markdown-it-plugin) or adding custom rulers (this is what the Live Preview component makes use of). Follows the CommonMark spec.

### [remarkable](https://github.com/jonschlinkert/remarkable)
The last one in the analysis is quite similar to markdown-it. Actually, the markdown-it developers are the authors of 99% of the Remarkable code, but they decided to create a new project (markdown-it) with new leadership. It's faster than markdown-it but has worst extensibility.

### Performance

I've created a small [jsperf test](https://jsperf.com/markdown-performance/1) suite to measure the speed of each library in the analysis.

### How markdown-it engine works and extensibility

The markdown-it engine has an ordered set of defined rules to parse each line in the Markdown source. The engine loops over each line, and [applies each rule sequentially](https://github.com/markdown-it/markdown-it/blob/master/lib/parser_block.js#L78) to the line of markdown source. If the rule is applied successfully, no more rules are evaluated over that line, and the engine keeps going with the next line. The API has methods to modify the ordered set of rules, so for example, a new paragraph rule could be evaluated before the existing one for all lines starting with upper case letters. This way you won't be replacing any existing rules.

The Live Preview component takes advantage of this to rewrite the `newline` and `paragraph` rules to keep heading and trailing spaces in each line. At first, I was reluctant to follow this path, but after checking the internals of the engine, I'm more happy with this approach. It follows the [API defined by the library](https://markdown-it.github.io/markdown-it/#Ruler.at), so we're not monkey-patching the library's engine.

### Conclussion

After having analyzed the most used Javascript Markdown parsers, I'm quite happy with the current approach: leveraging on the extensibility of the markdown-it parser. The performance is very good, the parser API is very flexible, and the current specification (CommonMark) can be extended easily with plugins to support Footnotes, definition lists, and some other niceties. The other options are either less performant, or less customizable, so I really don't see any other option than creating our own parser.

I'm pinging @dmsnell here because @ehg mentioned in Slack he could be interested in this.

Ferdev commented:
Apologies for the delay in my reply, @dmsnell! Our family is moving from Spain to the USA this August and I have a lot on my plate right now.

Thank you very much for showing your thoughts on this PR. I'm glad to see you are asking the same questions I asked myself when I started working on this. One of the things I missed after I checked the work already done on this feature, was some sort of documentation from a design standpoint. I was expecting those questions to be had been answered there. Another question I had was: where should be footnotes placed in a post with several different blocks? 

But in retrospective, I guess the most critical question that I asked myself and that I should've thrown in the open was: where does this functionality come from?

I guess I was a bit confused because I didn't see a definite place where all design questions and feedback had been gathered and discussed. Have you ever considered using a Request For Comments methodology, or do you use any other system to do so? I'd love to talk more about this.

I share your concern about the design problem this feature represents more than its coding counterpart. Actually, when I shared the first version with my assigned team, I asked for feedback from the designers more than anything.

I think it's a great idea to put together all my thoughts and decisions on a README :+1:. Just give me a couple of days to gather some time, and I'll push it to this PR.

Thank you again for your feedback and the time spent thinking about all this, Dennis! :bow:

brbrr commented:
Since our plan is to build JS (and maybe CSS) within Calypso env - we shouldn't merge this PR as is. JS/CSS part already merged here: https://github.com/Automattic/wp-calypso/pull/26336, the remainig part is to ship PHP code for this block. 

Moving this PR into next release.

simison commented:
Code for the block has been merged and lives in https://github.com/Automattic/wp-calypso/tree/1b8afd08cc4427389ce5632aec97e4e19136db31/client/gutenberg/extensions/markdown currently. We'll need to figure out PHP part of this separately in a different PR.

cc @tyxla @lezama
